services/provider/client/client.go

```diff
@@ -10,82 +10,103 @@ import (
 )
 
 type OCSProviderClient struct {
-	Client  pb.OCSProviderClient
-	timeout time.Duration
+	Client     pb.OCSProviderClient
+	clientConn *grpc.ClientConn
+	timeout    time.Duration
 }
 
-// NewProviderClient creates a client to talk to OCS provider server
-func NewProviderClient(cc *grpc.ClientConn, timeout time.Duration) *OCSProviderClient {
-	return &OCSProviderClient{Client: pb.NewOCSProviderClient(cc), timeout: timeout}
-}
-
-// NewGRPCConnection returns a grpc client connection which can be used to create the consumer client
-// Note: Close the connection after use
-func NewGRPCConnection(serverAddr string, timeout time.Duration) (*grpc.ClientConn, error) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+// NewProviderClient creates a client to talk to the external OCS storage provider server
+func NewProviderClient(ctx context.Context, serverAddr string, timeout time.Duration) (*OCSProviderClient, error) {
+	apiCtx, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
 	opts := []grpc.DialOption{
 		grpc.WithInsecure(),
 		grpc.WithBlock(),
 	}
 
-	conn, err := grpc.DialContext(ctx, serverAddr, opts...)
+	conn, err := grpc.DialContext(apiCtx, serverAddr, opts...)
 	if err != nil {
 		return nil, fmt.Errorf("failed to dial: %v", err)
 	}
 
-	return conn, err
+	return &OCSProviderClient{
+		Client:     pb.NewOCSProviderClient(conn),
+		clientConn: conn,
+		timeout:    timeout}, nil
+}
+
+// Close closes the gRPC connection of the external OCS storage provider client
+func (cc *OCSProviderClient) Close() {
+	_ = cc.clientConn.Close()
+	cc.clientConn = nil
+	cc.Client = nil
 }
 
 // OnboardConsumer to validate the consumer and create StorageConsumer
 // resource on the StorageProvider cluster
-func (cc *OCSProviderClient) OnboardConsumer(token, name, capacity string) (*pb.OnboardConsumerResponse, error) {
+func (cc *OCSProviderClient) OnboardConsumer(ctx context.Context, token, name, capacity string) (*pb.OnboardConsumerResponse, error) {
+	if cc.Client == nil || cc.clientConn == nil {
+		return nil, fmt.Errorf("provider client is closed")
+	}
+
```

Comment:
The entire construct is an abstraction on top of the JRPC client. 

In this specific abstraction, we only support a 1:1 relation between GRPC connections and clients and both of the objects share the same lifespan. The thing is that we want to release the underlying objects as soon as close() is called (to be garbage collected). So the check above just makes sure that none of the other methods are called after close() was called.

Reply:
Ah-ha, the garbage collection thing is something I didn't think of.... good enough. The only thing I'll say is that we should go a bit further and have something like an `isClosed()` method, since this is already repeated and would mirror the `Closed()` method. Still, no need to hold this PR for that, given the impending feature freeze.

Reply:
Agreed about the `isCLosed` observation. It will also allow us to expose that state of the client without exposing the implementation. We will take it in a future PR